File: _Easy/13/13.go
package main

import "fmt"

var symbols = map[rune]int{
	'I': 1,
	'V': 5,
	'X': 10,
	'L': 50,
	'C': 100,
	'D': 500,
	'M': 1000,
}

func romanToInt(s string) int {
	n := len(s)
	ans := 0
	for i := range s {
		val := symbols[rune(s[i])]
		if i < n-1 && val < symbols[rune(s[i+1])] {
			ans -= val
		} else {
			ans += val
		}
	}
	return ans
}

func main() {
	s := "VVVVVVVVVVVVVVVVV"
	ans := romanToInt(s)
	fmt.Printf("\n\n%d\n\n", ans)
}

File: _Easy/13/13.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
typedef pair<int, int> pii;
/* 
Roman numerals are represented by seven different symbols: I, V, X, L, C, D and M.

Symbol       Value
I             1
V             5
X             10
L             50
C             100
D             500
M             1000
For example, 2 is written as II in Roman numeral, just two ones added together. 12 is written as XII, which is simply X + II. The number 27 is written as XXVII, which is XX + V + II.

Roman numerals are usually written largest to smallest from left to right. However, the numeral for four is not IIII. Instead, the number four is written as IV. Because the one is before the five we subtract it making four. The same principle applies to the number nine, which is written as IX. There are six instances where subtraction is used:

I can be placed before V (5) and X (10) to make 4 and 9. 
X can be placed before L (50) and C (100) to make 40 and 90. 
C can be placed before D (500) and M (1000) to make 400 and 900.
Given a roman numeral, convert it to an integer.


Input: s = "III"
Output: 3
Explanation: III = 3.

Input: s = "LVIII"
Output: 58
Explanation: L = 50, V= 5, III = 3.

Input: s = "MCMXCIV"
Output: 1994
Explanation: M = 1000, CM = 900, XC = 90 and IV = 4.

方法一：遍历 + 哈希
每次加上
但如果 I 在 V 左边，则减去：即如果左边的值小于右边的值，则减去

 */
class Solution {
 public:
  int romanToInt(string s) {
    unordered_map<char, int> roman = {
        {'I', 1}, {'V', 5}, {'X', 10}, {'L', 50}, {'C', 100}, {'D', 500}, {'M', 1000},
    };
    int n = s.size();
    int ans = 0;
    for (int i = 0; i < n; i++) {
      if (i < n - 1 && roman[s[i]] < roman[s[i + 1]]) {
        ans -= roman[s[i]];
      } else {
        ans += roman[s[i]];
      }
    }
    return ans;
  }
};
int main() {

  Solution test;
  return 0;
}
